src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -504,10 +673,38 @@ private static MethodIL GenerateAccessor(ref GenerationContext context)
             // used to look up the target member to access and ignored
             // during dispatch.
             int beginIndex = context.IsTargetStatic ? 1 : 0;
-            int stubArgCount = context.Declaration.Signature.Length;
+            int stubArgCount = context.DeclarationSignature.Length;
+            Stubs.ILLocalVariable?[] localsToRestore = null;
             for (int i = beginIndex; i < stubArgCount; ++i)
             {
-                codeStream.EmitLdArg(i);
+                if (context.Declaration.Signature[i] != context.DeclarationSignature[i])
+                {
+                    if (context.DeclarationSignature[i] is { Category: TypeFlags.Class } classType)
+                    {
+                        codeStream.EmitLdArg(i);
+                        codeStream.Emit(ILOpcode.castclass, emit.NewToken(classType));
+                    }
+                    else if (context.DeclarationSignature[i] is ByRefType { ParameterType.Category: TypeFlags.Class } byrefType)
+                    {
+                        localsToRestore ??= new Stubs.ILLocalVariable?[stubArgCount];
+
+                        TypeDesc targetType = byrefType.ParameterType;
+                        codeStream.EmitLdArg(i);
+                        localsToRestore[i] = emit.NewLocal(targetType);
+                        codeStream.EmitLdInd(targetType);
+                        codeStream.Emit(ILOpcode.castclass, emit.NewToken(targetType));
```

Comment:
VM impl does unbox.any. This should do the same

https://github.com/dotnet/runtime/pull/114881/files#diff-c5fb5a7c28b2b8c88a86c638ea93d4419b5ca1a6d8fd500f1793c9a55df758f4R781